SDL/MiniDoublet.cuh

```diff
@@ -729,17 +719,114 @@ namespace SDL
 
                             addMDToMemory(mdsInGPU,hitsInGPU, modulesInGPU, lowerHitArrayIndex, upperHitArrayIndex, lowerModuleIndex, dz, dphi, dphichange, shiftedX, shiftedY, shiftedZ, noShiftedDz, noShiftedDphi, noShiftedDphiChange, mdIndex);
                         }
-
                     }
                 }
             }
         }
     };
 
-    
+    struct createMDArrayRangesGPU
+    {
+        template<typename TAcc>
+        ALPAKA_FN_ACC void operator()(
+                TAcc const & acc,
+                struct modules& modulesInGPU,
+                struct objectRanges& rangesInGPU) const
+        {
+            using Dim = alpaka::Dim<TAcc>;
+            using Idx = alpaka::Idx<TAcc>;
+            using Vec = alpaka::Vec<Dim, Idx>;
 
+            Vec const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+            Vec const gridThreadExtent = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
 
-}
+            // Initialize variables in shared memory and set to 0
+            int& nTotalMDs = alpaka::declareSharedVar<int, __COUNTER__>(acc); nTotalMDs = 0;
+            alpaka::syncBlockThreads(acc);
 
-#endif
+            // Initialize variables outside of the for loop.
+            float module_eta;
+            unsigned int nTotMDs;
+            int occupancy, category_number, eta_number;
+            short module_subdets, module_layers, module_rings;
```

Comment:
Since there is some moving around of the declaration of the variables, are we sure that we want them defined outside of the loop, even if they are not used anywhere else? I would think that the better practice would be to declare them inside of the scope they are used.
I also search around the internet and a lot of the selected replies in stackoverflow, etc. questions, e.g. [here](https://stackoverflow.com/questions/7959573/declaring-variables-inside-loops-good-practice-or-bad-practice), seem to agree that this could be better for compiler optimizations. I guess this is not very relevant here but it would be good to be consistent with our preferred practice.

Depending on which practice we decide to follow, this could imply changes to more places below.

Reply:
Thank you for looking into this question more. I think the comments about scope aren't relevant here because the for loop is wrapped in its own kernel already, so there's no chance of it causing issues downstream. However, since none of the others kernels do this, and there's probably no noticeable timing difference from it, I'll define them in the loop for now and we can keep this in the back of our minds.